controllers/gateway/apirule_controller.go

```diff
@@ -88,90 +72,155 @@ func (r *APIRuleReconciler) handleAPIRuleGetError(ctx context.Context, name type
-		if err != nil {
-			return doneReconcileErrorRequeue(r.OnErrorReconcilePeriod)
+	if err := r.Client.Get(ctx, req.NamespacedName, &apiRule); err != nil {
+		if apierrs.IsNotFound(err) {
+			// (Ressetkk): should be log.Debug...
+			//l.Info("ApiRule not found")
+			return doneReconcileNoRequeue()
 		}
-		cmd = r.getv2alpha1Reconciliation(apiRule, apiRuleV2alpha1, defaultDomainName, needsMigration, &namespacedLogger)
-	} else {
-		namespacedLogger.Info("Reconciling APIRule", "name", apiRule.Name, "namespace", apiRule.Namespace, "jwtHandler", r.Config.JWTHandler)
-		cmd = r.getV1beta1Reconciliation(apiRule, defaultDomainName, &namespacedLogger)
+		l.Error(err, "Error while getting APIRule")
+		return doneReconcileErrorRequeue(errorReconciliationPeriod)
 	}
 
-	if apiRuleErr != nil {
-		return r.handleAPIRuleGetError(ctx, req.NamespacedName, apiRule, apiRuleErr, cmd)
+	if !apiRule.DeletionTimestamp.IsZero() {
+		l.Info("APIRule is marked for deletion, deleting")
+		return r.reconcileAPIRuleDeletion(ctx, l, &apiRule)
 	}
 
-	shouldDeleteAPIRule := !apiRule.DeletionTimestamp.IsZero()
-	if !shouldDeleteAPIRule {
+	if !controllerutil.ContainsFinalizer(&apiRule, apiGatewayFinalizer) {
+		l.Info("APIRule is missing a finalizer, adding")
+		n := apiRule.DeepCopy() // copy to avoid conflicts
+		controllerutil.AddFinalizer(n, apiGatewayFinalizer)
+		// (Ressetkk): should be part of update logic to reduce number of
+		// updates on a resource
+		return r.updateResourceRequeue(ctx, l, n)
+	}
+
+	var cmd processing.ReconciliationCommand
+	if r.isApiRuleConvertedFromV2alpha1(apiRule) {
+		l.Info("Reconciling v2alpha1 APIRule")
+		migrate, err := apiRuleNeedsMigration(ctx, r.Client, &apiRule)
+		if err != nil {
+			return doneReconcileErrorRequeue(r.OnErrorReconcilePeriod)
+		}
+		if migrate {
+			n := apiRule.DeepCopy() // copy to avoid conflicts
+			migration.ApplyMigrationAnnotation(l, n)
+			// should not conflict with future updates as long as there are
+			// no Update() calls to the resource after that
+			if err := r.Client.Update(ctx, n); err != nil {
+				l.Error(err, "Failed to update migration annotation")
+				return doneReconcileErrorRequeue(r.OnErrorReconcilePeriod)
+			}
+		}
+		// grab latest v2alpha1
+		apiRuleV2alpha1 := gatewayv2alpha1.APIRule{}
+		if err := r.Client.Get(ctx, req.NamespacedName, &apiRuleV2alpha1); err != nil {
+			return doneReconcileErrorRequeue(r.OnErrorReconcilePeriod)
+		}
+		cmd = r.getv2alpha1Reconciliation(&apiRule, &apiRuleV2alpha1,
+			defaultDomainName, migrate, &l)
+
 		if name, err := dependencies.APIRule().AreAvailable(ctx, r.Client); err != nil {
-			apiRuleStatus, err := handleDependenciesError(name, err).ToAPIRuleStatus()
+			s, err := handleDependenciesError(name, err).ToAPIRuleStatus()
 			if err != nil {
 				return doneReconcileErrorRequeue(r.OnErrorReconcilePeriod)
 			}
-			return r.updateStatusOrRetry(ctx, apiRule, apiRuleStatus)
+			if err := s.UpdateStatus(&apiRuleV2alpha1.Status); err != nil {
```

Comment:
This Update will fail because `s` is of type `ReconciliationV1beta1Status` and therefore in `UpdateStatus` only `gatewayv1beta1.APIRuleStatus` is supported.